### DISCUSSION TITLE
[MTR05] Week1 超級挑戰題 github.sh

### DISCUSSION THREAD

**Discussion by benben6515** [Lidemy/forum#50] [程式導師實驗計畫] [answered]
超級挑戰題
#### 分析題目如下:
分別印出github user 的 name, bio, location, blog。
#### 目前做到 : 
```
#!/bin/bash

curl https://api.github.com/users/$1 | tail -n 15;
```

已知 curl 可以抓 API 的資料回來，使用 tail 抓尾巴我要的，
然後就是處理字串的問題。

#### Q : 無法使用變數紀錄 curl，Ex : 
```
#!/bin/bash

res=curl https://api.github.com/users/$1 | tail -n 15;
echo $res;
```
會Err...，找了好多資料好像也沒有這種用法
總不能 echo 一次就要 curl 一次吧

###### 感謝學長姐或是老師提示~

**Comment by aszx87410** [comment]
指令要用符號包住喔，有很多種寫法，例如說：

``` bash
#!/bin/bash

res=`curl https://api.github.com/users/$1 | tail -n 15`
echo $res;
```

我解答的過程：

1. 拿你的程式碼跑確實不行，看錯誤感覺是語法問題
2. google `bash variable result`
3. 點開搜尋結果第二篇：https://stackoverflow.com/questions/4651437/how-do-i-set-a-variable-to-the-output-of-a-command-in-bash

**Reply by benben6515** [reply]
謝謝老師的提示 ~ 
我想說我 `"` 跟 `'` 都是過了 原來是要用 ` 呀。
不過又卡關了 QQ
Linux 的字串處理太深奧啦~~~
可惡，要是 用 JS 寫一下就搞定了 >_<
最後放棄治療，寫了下面的糞code XDD

```
#!/bin/bash

res=`curl https://api.github.com/users/$1 | tail -n 15`;

echo $res | awk '{FS=":"} {print $2}';
echo $res | awk '{FS=":"} {print $6}';
echo $res | awk '{FS=":"} {print $8 " " $9}';
echo $res | awk '{FS=":"} {print $15 " " $16 " "$17 " "$18 " "$19" " $20 " "$21" " $22 $23 " "$24 " "$25 " "$26 " "$27 " "$28}';
```
![image](https://user-images.githubusercontent.com/61361198/114669043-0a346c00-9d34-11eb-8e12-690ef5a71f6e.png)

就這樣總共糞了 7 個小時 ...

**Comment by benben6515** [comment, marked as answer]
###### 2021 Apr. 15 update

老師解出來了！！！
雖然打了 4 次 API，但我覺得還可以接受 (應該吧)

會這樣做是因為如果先用變數存起來整個文章會變一個長長的 string ( 太神奇了吧 )，
我昨天就是這樣進入一個死胡同 Q_Q
因為只要一用 `grep` 全部都是跑整篇文章出來；
用 `cut` 也是一大串長 string 不知道怎麼 cut；
用 `awk` 也是跟 `cut` 一樣狀況。
昨天睡覺突然想到 : 其實打 4 次 API 也沒甚麼關係呀。Why not?

> 附圖為超級挑戰題的 code & output
![image](https://user-images.githubusercontent.com/61361198/114797574-09e4b100-9dc6-11eb-9c68-db4e11f8a524.png)

**Reply by aszx87410** [reply]
恭喜 🎉 
那既然都成功解出來了，就可以往後繼續優化了，第一個目標就是只打一次 API 就好XD

**Reply by JAS0NHUANG** [reply]
<details>
  <summary>給個關於「長長的 string 」的小提示，不想被爆雷不要點~~</summary>
  Quoting Variables
</details>

<details>
  <summary>另外關於那些 curl 顯示的資料……一樣不想被爆雷不要點:smile:</summary>
  silent
</details>

**Reply by benben6515** [reply]
###### 2021 Apr. 16 update

感謝老師的鼓勵跟助教的提示🙏🙏🙏

優化完成啦 🎉🎉🎉
有參考老師的解答，但優化後，只要符合 JSON 格式欄位怎麼變化都 OK ~

對初學者不太友善 Q_Q

要知道 JSON 、 Regular Expression(RE) 、gerp 的進階用法
好了，概念的部分差不多是這樣，以下詳解

<details >
<summary>!!! 防雷 !!!，我是詳解，附圖 & code</summary> 

![](https://i.imgur.com/3RPAXLO.png)

```
#!/bin/bash

res=`curl https://api.github.com/users/$1 -s | tail -n 15`;

echo $res | grep -o '"name": ".*",' | cut -d '"' -f4;
echo $res | grep -o '"bio": ".*",' | cut -d '"' -f4;
echo $res | grep -o '"location": ".*",' | cut -d '"' -f4;
echo $res | grep -o '"blog": ".*",' | cut -d '"' -f4;
```

`-s` 加上助教提示的參數，讓 output 乾淨很多
`tail -n 15` 取 JSON 倒數 15 行資料，優化後面匹配的效率
`gerp -o` 最關鍵的地方 ，`-o` 這參數平常根本找不到 ( 不用找了鳥哥那也沒有，我總共找了 8小時 ) 
`cut -d '"' -f4` 後面的處理簡單用 `cut` 就完成了 ( 可以參考鳥哥 cut 介紹)

</details>

**Reply by aszx87410** [reply]
我本來找參數也是用 google 去找，後來才發現其實直接 man grep 掃過每個選項其實也滿快的
